base/array.jl

```diff
@@ -1061,13 +1048,13 @@ julia> findnext(isodd, A, 2)
 0
 ```
 """
-function findnext(testf::Function, A, start::Integer)
-    for i = start:length(A)
-        if testf(A[i])
-            return i
-        end
+function findnext(testf::Function, A, idx::Integer)
+    while !done(A, idx)
+        lastidx = idx
+        elem, idx = next(A, idx)
+        testf(elem) && return lastidx
     end
-    return 0
```

Comment:
More importantly: you assume that `idx` is an `Int`, or the function will be type unstable. Need `zero(idx)`.